directives/meta.py

```diff
@@ -22,5 +35,63 @@ class AplusMeta(Directive):
         'introduction': directives.unchanged, # module introduction HTML
     }
 
+    # Valid date formats:
+    # 1. 'YYYY-MM-DD [hh[:mm[:ss]]]'
+    # 2. 'DD.MM.YYYY [hh[:mm[:ss]]]'
+    date_format = re.compile(r"^(\d\d\d\d-\d\d-\d\d|\d\d.\d\d.\d\d\d\d)( \d\d(:\d\d(:\d\d)?)?)?$")
+
+    # Keys in option_spec which require a date format
+    date_format_required = {
+        'open-time', 'read-open-time', 'close-time', 'late-time'
+    }
+
     def run(self):
-        return [aplusmeta(options=self.options)]
+        env = self.state.document.settings.env
+        aplusmeta_substitutions = env.config.aplusmeta_substitutions
+        modified_options = copy.deepcopy(self.options)
+
+        # Substitute values of options if a corresponding string is found in
+        # the configuration variable aplusmeta_substitutions (set in conf.py).
+        # Example:
+        #     self.options['open-time'] == 'open01'
+        #     aplusmeta_substitutions['open01'] == '2020-01-03 12:00'
+        #     # Result
+        #     modified_options['open-time'] = '2020-01-03 12:00'
+        #
+        # See the section "5. Meta (exercise round settings)" in README.md.
+        for opt, value in self.options.items():
+            if opt in AplusMeta.date_format_required:
+                self.validate_time(opt, value, aplusmeta_substitutions)
+            if value in aplusmeta_substitutions:
+                modified_options[opt] = aplusmeta_substitutions[value]
+        return [aplusmeta(options=modified_options)]
+
+    def validate_time(self, opt, value, aplusmeta_substitutions):
+        # Validates the time of given option-value pair. Also checks possible
+        # substitutions in the configuration variable 'aplusmeta_substitutions'.
+        # Raises a SphinxError if value is invalid (even with substitution).
+        source, line = self.state_machine.get_source_and_line(self.lineno)
+        position_text = ("{}, line {}, directive aplusmeta:\n" +
+                         "option '{}' has value '{}' ")
+        date_text = ("1. Date in format 'YYYY-MM-DD [hh[:mm[:ss]]]', e.g. " +
+                    "'2020-01-16 16:00'\n" +
+                    "2. Date in format 'DD.MM.YYYY [hh[:mm[:ss]]]', e.g. " +
+                    "'16.01.2020 16:00'\n")
+        is_time = AplusMeta.date_format.match
+        if not is_time(value):
+            if value not in aplusmeta_substitutions:
+                raise SphinxError((position_text +
+                    "which was not recognised.\n" +
+                    "This should be one of:\n" + date_text +
```

Comment:
This is slower (most likely) than just writing both template strings separately. This also makes the code harder to read as you need to concatenate the strings in your head. Just write both strings with placeholders, so they are easy to read and fix in the code.

Reply:
Very good, I'll do this.